crates/networking/manager/src/gossipsub/validate/sync_committee_contribution_and_proof.rs

```diff
@@ -0,0 +1,230 @@
+    constants::beacon::{DOMAIN_SYNC_COMMITTEE, SYNC_COMMITTEE_SIZE},
+    misc::{compute_epoch_at_slot, compute_signing_root, compute_sync_committee_period},
+};
+use ream_storage::{
+    cache::{CacheSyncCommitteeContribution, CachedDB, SyncCommitteeKey},
+    tables::table::Table,
+};
+use ream_validator_beacon::{
+    constants::{
+        DOMAIN_CONTRIBUTION_AND_PROOF, DOMAIN_SYNC_COMMITTEE_SELECTION_PROOF,
+        SYNC_COMMITTEE_SUBNET_COUNT,
+    },
+    contribution_and_proof::SignedContributionAndProof,
+    sync_committee::{SyncAggregatorSelectionData, is_sync_committee_aggregator},
+};
+
+use super::result::ValidationResult;
+
+pub async fn validate_sync_committee_contribution_and_proof(
+    beacon_chain: &BeaconChain,
+    cached_db: &CachedDB,
+    signed_contribution_and_proof: &SignedContributionAndProof,
+) -> anyhow::Result<ValidationResult> {
+    let contribution_and_proof = &signed_contribution_and_proof.message;
+    let contribution = &contribution_and_proof.contribution;
+
+    let store = beacon_chain.store.lock().await;
+    let head_root = store.get_head()?;
+
+    let block = store
+        .db
+        .beacon_block_provider()
+        .get(head_root)?
+        .ok_or_else(|| anyhow!("Could not get block for head root: {head_root}"))?;
+
+    let state = store
+        .db
+        .beacon_state_provider()
+        .get(head_root)?
+        .ok_or_else(|| anyhow!("No beacon state found for head root: {head_root}"))?;
+
+    let current_slot: u64 = block.message.slot;
+
+    // [IGNORE] if contribution.slot is equal to or earlier than the current_slot (with a
+    // MAXIMUM_GOSSIP_CLOCK_DISPARITY allowance)
+    if contribution.slot != current_slot {
+        return Ok(ValidationResult::Ignore(
+            "Contribution is from a future slot".to_string(),
+        ));
+    }
+
+    // [REJECT] if contribution.subcommittee_index is out of SYNC_COMMITTEE_SUBNET_COUNT range.
+    if contribution.subcommittee_index >= SYNC_COMMITTEE_SUBNET_COUNT {
+        return Ok(ValidationResult::Reject(
+            "The subcommittee index is out of range".to_string(),
+        ));
+    }
+
+    // [REJECT] if contribution doesn't have any participants
+    if contribution.aggregation_bits.num_set_bits() == 0 {
+        return Ok(ValidationResult::Reject(
+            "The contribution has too many participants".to_string(),
+        ));
+    }
+
+    // [REJECT] if is_sync_committee_aggregator(contribution_and_proof.selection_proof) is false
+    if !is_sync_committee_aggregator(&contribution_and_proof.selection_proof) {
+        return Ok(ValidationResult::Reject(
+            "The selection proof is not a valid aggregator".to_string(),
+        ));
+    }
+
+    // [REJECT] if the validator with index contribution_and_proof.aggregator_index is not in the
+    // sync committee for the epoch of contribution.slot
+    let validator_pubkey =
+        &state.validators[contribution_and_proof.aggregator_index as usize].public_key;
```

Comment:
we shouldn't be doing raw parses in code which runs remote payloads, as they could cause us to panic